### ISSUE TITLE
Coq master behavioral change: eexists leaves more evars, breaking ssreflect rewrite

### ISSUE THREAD

**Issue by RalfJung** [coq/coq#7936]
#### Version

master (b98ae49d)

#### Operating system

Linux

#### Description of the problem

Consider the following example (took forever to minimize...)

```coq
Require Export Utf8 List Setoid Program.

Global Generalizable All Variables.

Delimit Scope stdpp_scope with stdpp.
Global Open Scope stdpp_scope.
Notation "(=@{ A } )" := (@eq A) (only parsing) : stdpp_scope.

Class Decision (P : Prop) := decide : {P} + {¬P}.
Arguments decide _ {_} : simpl never, assert.

Class RelDecision {A B} (R : A → B → Prop) :=
  decide_rel x y :> Decision (R x y).
Notation EqDecision A := (RelDecision (=@{A})).

Definition strict {A} (R : relation A) : relation A := λ X Y, R X Y ∧ ¬R Y X.

Infix "∘" := compose : stdpp_scope.

Class SubsetEq A := subseteq: relation A.
Notation "(⊆)" := subseteq (only parsing) : stdpp_scope.
Notation "( X ⊂)" := (strict (⊆) X) (only parsing) : stdpp_scope.

Class Filter A B := filter: ∀ (P : A → Prop) `{∀ x, Decision (P x)}, B → B.
Class FMap (M : Type → Type) := fmap : ∀ {A B}, (A → B) → M A → M B.

Infix "<$>" := fmap (at level 61, left associativity) : stdpp_scope.

Class Lookup (K A M : Type) := lookup: K → M → option A.
Notation "m !! i" := (lookup i m) (at level 20) : stdpp_scope.

Class Insert (K A M : Type) := insert: K → A → M → M.
Notation "<[ k := a ]>" := (insert k a)
  (at level 5, right associativity, format "<[ k := a ]>") : stdpp_scope.

Class Elements A C := elements: C → list A.

Class SqSubsetEq A := sqsubseteq: relation A.

Class Join A := join: A → A → A.
Infix "⊔" := join (at level 50) : stdpp_scope.
Notation "(⊔)" := join (only parsing) : stdpp_scope.

Definition from_option {A B} (f : A → B) (y : B) (mx : option A) : B :=
  match mx with None => y | Some x => f x end.

Module Export functions.

Section definitions.
  Context {A T : Type} `{EqDecision A}.
  Global Instance fn_insert : Insert A T (A → T) :=
    λ a t f b, if decide (a = b) then t else f b.
End definitions.

Section functions.
  Context {A T : Type} `{!EqDecision A}.

  Lemma fn_lookup_insert (f : A → T) a t : <[a:=t]>f a = t.
  Admitted.
End functions.
Require Export ssreflect.

Lemma test (Lat : Type) `{!SqSubsetEq Lat, !Join Lat}
      `{Lookup nat (bool * Lat) MAP} `{!EqDecision nat} `{Elements SET nat}
      (Vκ : Lat) (κ : nat) (A : MAP) (Vs : nat → Lat) (K : SET) : True.
assert (∃ Vs : nat → Lat, ∀ Λ : nat, from_option (sqsubseteq (Vs κ) ∘ snd) False (A !! Λ)).
eexists (<[κ := Vκ ⊔ fold_right (⊔) Vκ (Vs <$> elements (filter (κ ⊂) K))]>Vs).
intros.
Set Printing All.
rewrite functions.fn_lookup_insert.
Abort.
```

This compiles fine with Coq 8.8.0, but breaks with master saying
```
Ltac call to "rewrite (ssrrwargs) (ssrclauses)" failed.
The LHS of (@fn_lookup_insert _ _ _)
    (@insert _ _ (forall _ : _, _) (@fn_insert _ _ _) _ _ _ _)
does not match any subterm of the goal
```

The reason is that `eexists` produces a different term. In 8.8.0, the goal after `eexists` is
```
  @from_option (prod bool Lat) Prop
    (@compose (prod bool Lat) Lat Prop
       (@sqsubseteq Lat SqSubsetEq0
          (@insert nat Lat (forall _ : nat, Lat) (@fn_insert nat Lat EqDecision0) κ
             (@join Lat Join0 Vκ
                (@fold_right Lat Lat (@join Lat Join0) Vκ
                   (@fmap list ?FMap nat Lat Vs
                      (@elements nat SET ?Elements
                         (@filter nat SET ?Filter (@strict nat (@subseteq nat ?SubsetEq) κ) ?H K)))))
             Vs κ)) (@snd bool Lat)) False (@lookup nat (prod bool Lat) MAP H Λ A)
```
whereas in master, it is
```
  @from_option (prod bool Lat) Prop
    (@compose (prod bool Lat) Lat Prop
       (@sqsubseteq Lat SqSubsetEq0
          (@insert nat Lat (forall _ : nat, Lat) ?Insert κ
             (@join Lat ?Join Vκ
                (@fold_right Lat Lat (@join Lat ?Join0) Vκ
                   (@fmap list ?FMap nat Lat Vs
                      (@elements nat SET ?Elements
                         (@filter nat SET ?Filter (@strict nat (@subseteq nat ?SubsetEq) κ) ?H K)))))
             Vs κ)) (@snd bool Lat)) False (@lookup nat (prod bool Lat) MAP H Λ A)
```
As you can see, there are more evars.

I am not sure if that's a bug, but `rewrite` failing where `rewrite ->` works is certainly a bug I would say.

(If you want to see this in action, [here](https://gitlab.mpi-sws.org/FP/LambdaRust-coq/commit/32684d0bd38b91ecc42f2862bcf9c0d65e294179) is the patch in a real development that was necessary to make things compile with Coq master again.)

**Comment by RalfJung** [comment]
Bisecting says

8d8200d4bff3ffc44efc51ad44dccee9eb14ec6a is the first bad commit
Author: Matthieu Sozeau <mattam@mattam.org>
Date:   Thu Jul 27 20:32:18 2017 +0200

>    Do not allow spliting in res_pf, this is reserved for pretyping

Cc @mattam82

**Comment by RalfJung** [comment]
I verified that reverting 8d8200d4bff3ffc44efc51ad44dccee9eb14ec6a fixes the testcase.

**Comment by mattam82** [comment]
Curious bug indeed. Is it really expected that the Join instance can be resolved while the FMap/Elements/SubsetEq instances cannot in this use case?

**Comment by Zimmi48** [comment]
I'm glad I decided not to backport this commit!

**Comment by RalfJung** [comment]
> Is it really expected that the Join instance can be resolved while the FMap/Elements/SubsetEq instances cannot in this use case?

I have no idea. I don't know why it resolves some instances but leaves others untouched.
Notice that
```
unshelve eexists (<[κ := Vκ ⊔ fold_right (⊔) Vκ (Vs <$> elements (filter (κ ⊂) K))]>Vs); try apply _; [shelve..|].
```
Solves the `Join` instance even with Coq master. Just `eexists` itself doesn't do it any more.

---

Also @gares is it expected that ssreflect's `rewrite` fails here even though Coq's `rewrite ->` succeeds?

**Comment by silene** [comment]
@mattam82 Should I revert the commit on 8.9? I have a hard time evaluating the impact here.

**Comment by mattam82** [comment]
Yes you can revert. @RalfJung it resolved that instance and not the others because they are independent: the Join constraint does not share existentials with the other ones I suppose. Your script using try apply _ reproduces that behavior. The goal of splitting like this was to provide more precise error messages: only the typeclass constraint components that contain a constraint that cannot be solved would be displayed. In a fail:false context this makes a bit less sense. I think I had a more compelling reason to disallow it in res_pf coming from a use-case in the unifall branch but can’t remember it now. For sure it will come back to me when I make progress on that branch.

**Comment by mattam82** [comment]
I suppose ssreflect’s rewrite doesn’t use split: either, or uses fail:true

**Comment by gares** [comment]
This bug is fixed in v8.9 but not in master. How should we proceed? Forward port the revert waiting for a better fix? Change milestone to 8.10?

**Comment by mattam82** [comment]
Yes, for the time being a revert is appropriate.
